### DISCUSSION TITLE
Define Quarkus Properties from env vars

### DISCUSSION THREAD

**Discussion by 0xdbe** [keycloak/keycloak#22798] [Archive]
Hi!

This [UserStorage](https://github.com/keycloak/keycloak-quickstarts/tree/release/22.0/extension/user-storage-jpa) provider in Keycloak Quickstarts, use a Quarkus DataSource.

Are there any way to set Quarkus properties without to create  `conf/quarkus.properties` file ? 
Create a file is not suitable for deployment in a Kubernetes cluster.

I tried to used env vars without success:

```
export QUARKUS_DATASOURCE_<DS_NAME>_JDBC_URL=jdbc:postgresql://localhost:5432/<DB_NAME>
```

Are there any other options ?

**Comment by 0xdbe** [comment]
I tried many options to set a Quarkus Datasource via env vars (QUARKUS_DATASOURCE*, JAVA_OPTS_APPEND, ...).
But it does not work, the variables are not taken into consideration by the keycloak.

Next step: try to set ` quarkus.properties` in a ConfigMap and mount it as a file.

**Reply by justin-tay** [reply]
Those particular properties are fixed at build time https://quarkus.io/guides/all-config#quarkus-datasource_quarkus.datasource.data-sources-datasources so make sure a `build` step gets triggered https://www.keycloak.org/server/containers.

**Reply by 0xdbe** [reply]
Do you know if these properties must be set  **before building provider** or **before starting keycloak** ?

since I use `start-dev` to start keycloak, I think build step is always run ?
Maybe if there is no change in provider, properties are not reloaded.

**Reply by 0xdbe** [reply]
I tried this:
- update and deploy a new version of the provider,
- run keycloak with env var `QUARKUS_DATASOURCE_<...>=**** kc.sh start-dev`

same error:

```
Could not obtain connection to query metadata: 
java.lang.UnsupportedOperationException: 
The application must supply JDBC connections
```

Workaround: use a `quarkus.properties` with env vars

```
quarkus.datasource.<NAME>.db-kind=postgresql 
quarkus.datasource.<NAME>.username=${DATABASE_USER_NAME:"keycloak"}
quarkus.datasource.<NAME>.password=${DATABASE_USER_PWD:"pAssw0rD"}
quarkus.datasource.<NAME>.jdbc.url=${DATABASE_JDBC_URL:"jdbc:postgresql://localhost:5432/<DB_NAME>"}
quarkus.datasource.<NAME>.jdbc.max-size=16
```

**Reply by kikkauz** [reply]
@0xdbe thanks for the workaround, I was looking exactly how to solve this issue. 
I have to build a docker image of keycloak that must be deployed in different environment, each one with its own DB instance, and I don't want to have an image per environment with sensible data in it!
